benchexec/container.py

```diff
@@ -531,29 +568,51 @@ def duplicate_mount_hierarchy(mount_base, temp_base, work_base, dir_modes):
             work_path = work_base + b"/" + str(overlay_count).encode()
             os.makedirs(temp_path, exist_ok=True)
             os.makedirs(work_path, exist_ok=True)
-            try:
-                # Previous mount in this place not needed if replaced with overlay dir.
-                libc.umount(mount_path)
-            except OSError as e:
-                logging.debug(e)
-            try:
-                make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
-            except OSError as e:
-                mp = mountpoint.decode()
-                raise OSError(
-                    e.errno,
-                    f"Creating overlay mount for '{mp}' failed: {os.strerror(e.errno)}. "
-                    f"Please use other directory modes, "
-                    f"for example '--read-only-dir {shlex.quote(mp)}'.",
-                )
+            if os.path.ismount(mount_path):
+                try:
+                    # Previous mount in this place not needed if replaced with overlay dir.
+                    libc.umount(mount_path)
+                except OSError as e:
+                    logging.debug(e)
+            if use_fuse:
+                fuse_mount_path = temp_fuse + mountpoint
+                make_bind_mount(fuse_mount_path, mount_path)
+            else:
+                try:
+                    make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
+                except OSError as e:
+                    # Resort to fuse-overlayfs if kernel overlayfs is not available.
```

Comment:
I think it should be doable to also use the "only one fuse mount" strategy here, right?

If we end up here, we can create the global fuse mount and then the bind mount as in the `use_fuse` case. We just need to remember in a variable whether the global fuse mount already exists such that we do not create it twice.

Reply:
plz review this:

https://github.com/sosy-lab/benchexec/blob/195e4d06c3db6b042ec80239acada55927a27d7f/benchexec/container.py#L542-L580

Reply:
This switches from kernel overlayfs to fuse-overlayfs as soon as the first mount problem of kernel overlayfs occurs. It makes it somewhat more consistent in the sense that we use fuse-overlayfs more often, but it also makes it really hard to predict whether kernel overlayfs or fuse-overlayfs is used for a specific mountpoint, as it depends on the order of mountpoints.

I think I would prefer if we change that to use fuse-overlayfs only for those mounts where it is necessary. Then at least it is easier to understand for every mountpoint what mode it has. Of course the case that we detect upfront that we need fuse-overlayfs should remain.

We just need a second boolean `fuse_overlay_mount_available` or so in addition to `use_fuse`.

Reply:
I think we can initialize the variable `fuse` as `None` and use it to indicate whether `fuse-overlayfs` is available, then it's like:

```python
            if use_fuse:
                fuse_mount_path = temp_base + b"/fuse" + mountpoint
                make_bind_mount(fuse_mount_path, mount_path)
            else:
                try:
                    make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
                except OSError as e:
                    # Resort to fuse-overlayfs if kernel overlayfs is not available.
                    # This part of the code (using fuse-overlayfs as a fallback) is intentionally
                    # kept as a workaround for triple-nested execution with kernel overlayfs.
                    if fuse is None:
                        fuse = shutil.which("fuse-overlayfs")
                        if fuse:
                            fuse_overlay_mount_root(fuse, temp_base, work_base)
                        else:
                            mp = mountpoint.decode()
                            logging.warning("fuse-overlayfs is not available.")
                            raise OSError(
                                e.errno,
                                f"Failed to create overlay mount for '{mp}': {os.strerror(e.errno)}. "
                                f"Consider using alternative directory modes, such as '--read-only-dir {shlex.quote(mp)}'.",
                            ) from e

                    logging.debug("Fallback to fuse-overlayfs for overlay mount.")
                    fuse_mount_path = f"{temp_base}/fuse{mountpoint.decode()}"
                    make_bind_mount(fuse_mount_path, mount_path)
```

Reply:
Yes, that would work, but it also kind of hides the fact that the "root" fuse-overlayfs mount was already made.

And actually, the path to the `fuse-overlayfs` binary is never needed again after `fuse_overlay_mount_root` is called. So I would instead say that all of the setup should be moved into `fuse_overlay_mount_root`, i.e., including searching for the executable. The function could be called something like `setup_fuse_overlay` or so and return an indicator of whether it was successful. This would also help in making the main function shorter.

Ideally the fallback code here would be only something like:
```python
if not fuse_overlay_mount_path:
    fuse_overlay_mount_path = setup_fuse_overlay(...)
    if not fuse_overlay_mount_path:
         # error, fuse-overlayfs not available
make_bind_mount(fuse_mount_path + mountpoint, mount_path)
```

We should not even have two places where `b"/fuse"` is used. It is always better to avoid having to write down such magic constants twice, to guard against future changes.

Reply:
Looks good now!